sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionBasedLoadBalancer.java

```diff
@@ -286,10 +286,101 @@ private Mono<Void> loadBalance(final Tuple2<Map<String, PartitionOwnership>, Lis
+         */
+        if (!CoreUtils.isNullOrEmpty(partitionsCache.get())) {
+            return Mono.empty();
+        }
+
+        /*
+         * Retrieve current partition ownership details from the datastore.
+         */
+        final Mono<Map<String, PartitionOwnership>> partitionOwnershipMono = checkpointStore
+            .listOwnership(fullyQualifiedNamespace, eventHubName, consumerGroupName)
+            .timeout(Duration.ofMinutes(1))
+            .collectMap(PartitionOwnership::getPartitionId, Function.identity());
+
+        Mono<List<String>> partitionsMono = eventHubAsyncClient
+            .getPartitionIds()
+            .timeout(Duration.ofMinutes(1))
+            .collectList();
+
+        return Mono.zip(partitionOwnershipMono, partitionsMono)
+            .flatMap((Tuple2<Map<String, PartitionOwnership>, List<String>> tuple) -> {
+                Map<String, PartitionOwnership> partitionOwnershipMap = tuple.getT1();
+
+                List<String> partitionIds = tuple.getT2();
+
+                if (CoreUtils.isNullOrEmpty(partitionIds)) {
+                    // This may be due to an error when getting Event Hub metadata.
+                    throw logger.logExceptionAsError(Exceptions.propagate(
+                        new IllegalStateException("There are no partitions in Event Hub " + eventHubName)));
+                }
+                partitionsCache.set(partitionIds);
+
+                if (!isValid(partitionOwnershipMap)) {
+                    // User data is corrupt.
+                    throw logger.logExceptionAsError(Exceptions.propagate(
+                        new IllegalStateException("Invalid partitionOwnership data from CheckpointStore")));
+                }
+
+                /*
+                 * Remove all partitions' ownership that have not been modified for a configuration period of time. This
+                 * means that the previous EventProcessor that owned the partition is probably down and the partition is now
+                 * eligible to be claimed by other EventProcessors.
+                 */
+                Map<String, PartitionOwnership> activePartitionOwnershipMap = removeInactivePartitionOwnerships(
+                    partitionOwnershipMap);
+
+                /*
+                 * If there are active partition ownership, this means the checkpoint store claimed partition successfully,
+                 * we don't need to check it.
+                 */
+                if (!CoreUtils.isNullOrEmpty(activePartitionOwnershipMap)) {
+                    return Mono.empty();
+                }
+
+                /*
+                 * If there is no active partition ownership, claim a random partition and the other partitions will be
+                 * claimed in load balance.
+                 */
+                PartitionOwnership ownershipRequest = createPartitionOwnershipRequest(partitionOwnershipMap,
+                    partitionIds.get(RANDOM.nextInt(partitionIds.size())));
```

Comment:
Does RANDON better than just take first partition?

Reply:
I'm not sure. But in loadbalance, if there are no partition start pump, it take a random partition to start as well. I keep the same logic.